[config] add `extra_tags` setting

vickenty opened the pull request:
### What does this PR do?

`extra_tags` specifies additional tags that will be appended to the host tags. Allows tags to be set via both the configuration file and environment variable at the same time. Duplicates will be removed.

### Describe your test plan

Provide some tags in `datadog.yaml` and set `DD_EXTRA_TAGS` to a different set of tags. Verify that tags from both sources are being sent.

vickenty committed:
[config] add new `extra_tags` setting

`extra_tags` specifies additional tags that will be appended to the host
tags. Allows tags to be set via both the configuration file and
environment variable at the same time.

vickenty committed:
[config] deduplicate host tags

Addition of `extra_tags` introduces a chance that duplicate host tags
will be specified. Remove duplicates to reduce network load.

vickenty committed:
[config] apply extra_tags everywhere

Refactor the code, so that the `extra_tags` setting is applied
everywhere, where `tags` was used.

ogaca-dd commented on pkg/metadata/host/host_tags.go:
```diff
@@ -145,7 +145,7 @@ func GetHostTags(cached bool) *Tags {
 
 	t := &Tags{
-		System:              hostTags,
+		System:              util.SortUniqInPlace(hostTags),
```
What do you think about moving the call `util.SortUniqInPlace` inside `GetConfiguredTags`?

vickenty replied:
This way de-duplication would only apply to configured tags before splitting, while duplicates can still occur due to tag splitting via `tag_value_split_separator`, and by having redundant or overlapping tags from k8s/gke/ec2.

ogaca-dd reviewed (commented):
LGTM!

Note: I think it worth adding a section inside `pkg/config/config_template.yaml`.

vickenty committed:
[config] add `extra_tags` to config template

vickenty commented:
> Note: I think it worth adding a section inside `pkg/config/config_template.yaml`.

Added. Any other documentation that needs updating?

sarina-dd reviewed (commented):
👍  LGTM release notes

clamoriniere commented:
Hi @vickenty 
I think this feature is useful, but maybe we should to it a bit differently.
My concerne is: the new `extra_tags` can also be part of the configuration. so at one point someone will ask for `extra_extra_tags` because they are using `tags` and `extra_tags` with two different deployment template tools, and so you will need another extra variable for providing tags coming from env var.

Maybe `extra_tags` should only be configurable with an env-var and not with the configuration file. And be more clear `extra_tags` could be rename `extra_tags_from_env`.
WDYT?

clamoriniere reviewed (commented):
Hi @vickenty 
I think this feature is useful, but maybe we should to it a bit differently.
My concerne is: the new `extra_tags` can also be part of the configuration. so at one point someone will ask for `extra_extra_tags` because they are using `tags` and `extra_tags` with two different deployment template tools, and so you will need another extra variable for providing tags coming from env var.

Maybe `extra_tags` should only be configurable with an env-var and not with the configuration file. And be more clear `extra_tags` could be rename `extra_tags_from_env`.
WDYT?

lucaspimentel commented on pkg/config/config_template.yaml:
```diff
@@ -67,6 +67,16 @@ api_key:
 #   - <TAG_KEY>:<TAG_VALUE>
 
+## @param extra_tags - list of key:value elements - optional
+## List of aditional host tqgs. Attached in-app to every metric, event, log,
```
```suggestion
## List of aditional host tags. Attached in-app to every metric, event, log,
```

vickenty commented:
> My concerne is: the new `extra_tags` can also be part of the configuration. so at one point someone will ask for `extra_extra_tags` because they are using `tags` and `extra_tags` with two different deployment template tools, and so you will need another extra variable for providing tags coming from env var.

Interesting point, I can definitely see some confusion here. I was thinking that customers that use `DD_TAGS` today would be able to add tags via configuration file without having to update their deployments, but I don't have a strong opinion here. Out of curiosity, is it possible to use two different deployment tools to set tags in the currently released agent?

vickenty committed:
Update pkg/config/config_template.yaml

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

vickenty committed:
[config] change how `extra_tags` is explained

Avoid potential confusion by only mentioning `extra_tags` in its
environment variable form, which is in line with the primary use case of
supplementing tags in the configuration file. But we still want users
that want to use it in the configuration file to be able to do so.

vickenty commented:
@clamoriniere I discussed this with the team, and we still want this to be available as a configuration file option, even if undocumented. I have removed `extra_tags` from the config template to avoid possible confusion, and instead it is mentioned as an environment variable in the `tags` comment. Does this address your concern?

clamoriniere reviewed (approved):
Thanks for the update. 
It looks ok to me.
